docs/06-contributors/999-rfcs/2022-04-30-preflight-instance-id.md

```diff
@@ -0,0 +1,70 @@
+---
+title: (Draft) Unique Preflight Instances ID
+description: How to create unique IDs for preflight instances
+---
+- **Author(s):**: @ekeren
+- **Submission Date**: {2022-04-30}
+- **Stage**: Draft
+
+## Background
+
+In the current implementation of Wing, when instantiating a new preflight class, the compiler provides a default ID for the instance. 
+In cases where the default ID cannot be made unique, a compiler error occurs.
+
+In the last wingathon, more than 70% of preflight instances in the Wing research repository are created with a custom ID. 
+
+## Observed Patterns
+
+When examining the code, the following patterns can be observed:
+
+1. Improved readability of preflight instances, e.g.:
+  
+    [dogfooding/publications-etl](https://github.com/winglang/research/blob/926453f5c5abd2e3158b4c8c659f1e790c99d741//app.w#L70):
+     ```ts (wing)
+     this.initialized = new cloud.Counter() as "initialized";
+     ```
+2. Name duplications - situations where the ID (or part of it) is also reflected in the constructor:
+  
+    [dogfooding/release-webhool](https://github.com/winglang/research/blob/926453f5c5abd2e3158b4c8c659f1e790c99d741/dogfooding/release-webhook/main.w#L207):
+     ```ts (wing)
+     let slack_token = new cloud.Secret(name: "slack-token") as "Slack Token";
+     ```
+     You can also consider situations where `cloud.Bucket` and `cloud.Test` where the ID of the instance and the name of bucket/test are connected 
+3. Multiple instances of the same type in the same scope:
+   
+     [dogfooding/publications-etl](https://github.com/winglang/research/blob/926453f5c5abd2e3158b4c8c659f1e790c99d741/dogfooding/publications-etl/app.w#LL31-L33C58):
+     ```ts (wing)
+     this.inbox = new cloud.Bucket() as "inbox";
+     this.transformations = new cloud.Bucket() as "transformations";
+     this.catalogs = new cloud.Bucket() as "xslt catalog";
+     ```
+
+## Proposal
+
+Differentiate between stateful instance (cloud.Bucket, Redis, cloud.Counter, cloud.Table, cloud.Queue) that require a stable ID, 
+and stateless instances where the ID is essential for discoverability and debuggability.
+
+- For stateful instances, provide the instance name as the first argument:
+  ```ts (wing)
+  new cloud.Bucket("inbox");
+  this.ingestion = new cloud.Queue("ingestion work queue");
+  new cloud.Test("bucket get and test", ...);
+  ```
+  
+- For stateless instances, default the instance name to the corresponding variable:
+  ```ts (wing)
+  // ./app.w
+  bring cloud; 
+  let my_func = new cloud.Function(...); // named "my_func"
+  new cloud.Function(...); // when no variable, default to named "cloud.Function$1"
+  new cloud.Function(...); // when no variable, default to named "cloud.Function$2"
+  ```
```

Comment:
@eladb , do you see a problem with not having a stable id for non persistent resources?

Reply:
Yes, stable IDs are important even for non-persistent resources, mainly due to the fact that infrastructure is (sometimes) hard to provision and there could be "in-flight" transactions that you don't want to disrupt.

Take a VPC for example - it's a non persistent resource but you definitely don't want it to be recreated if it's identity changes unexpectedly.

Even a Lambda function, you don't want to recreated it unnecessarily - it would take longer to deploy, logs will get deleted (sort of persistent things), etc.

If think the low hanging fruit would be the cosmetic/ergonomic change to use the variable name as display name (if defined).

We *could* explore the idea of requiring an `id` for persistent resources like `cloud.Bucket` and `cloud.Table` by simply adding a positional argument:

```js
class Bucket {
  constructor(scope: Construct, id: string, actualId: string, props: BucketProps) {
    super(scope, actualId);

    // ...
  }
}
```

This means that when you define a `cloud.Bucket` in Wing, you'll do it like this:

```js
new cloud.Bucket("my-bucket");
```

Note that if you use `as "foo"` it will be ignored (we can add some special casing to the compiler to disallow of course).

Maybe we can try it out for buckets and see how it feels and then make a decision.

Reply:
This is great feedback, I'll work to add this information to the RFC.